### ISSUE TITLE
Release version with the latest changes

### ISSUE THREAD

**Issue by rajatkapoor** [inukshuk/anystyle-parser#86]

**Comment by inukshuk** [comment]
We're currently re-writing and improving large parts of the parser and model. Will be released when it's ready as 1.0.0.

**Comment by inukshuk** [comment]
We released 1.0 a couple of months ago; there are still features that I am working on, but I'll be careful not to break backwards compatibility from now on. You can check the [HISTORY](https://github.com/inukshuk/anystyle/blob/master/HISTORY.md) file for a summary of recent changes. Most notably, we added the `Finder` component for extracting references from full-texts, however, this should be considered experimental at this stage (the data set is currently limited to theses and publications from a single university). The `Parser` component has been much improved over the 0.x version and also has much better tooling now. I would consider the parser component as pretty stable at this point, it is just mostly lacking documentation.

How does it compare to ParsCit and FreeCite? I have not compared results, but I would be very interested if someone did compare them. AnyStyle is definitely indebted to both these projects, but I started it mostly because I was not getting satisfying results for my own data sets. The parser component's architecture is based on the original ParsCit paper, but using Ruby + Wapiti instead of Perl + CRFPP. Parts of the feature extraction code of the 0.x version was influenced by FreeCite, but I don't think there is much resemblance left in the 1.0 release. Comparing results is probably not straight forward because each parser will be trained using slightly different labels, but if you want to give it a shot our [gold](https://github.com/inukshuk/anystyle/blob/master/res/parser/gold.xml) and [core](https://github.com/inukshuk/anystyle/blob/master/res/parser/core.xml) data sets comprise well over 2 thousand curated references; core includes a heavily modified version of the CORA set used by ParsCit (modified because the parser has detected numerous inconsistencies and errors over time). If you decide to run comparisons, please use the 1.0 version of AnyStyle and the default model (anystyle.io is still based on the 0.x version and the model can be trained by anyone, therefore its quality can vary greatly).

As for a long-term perspective, AnyStyle has mostly been a side project for me (the finder component was contract work though), and, while I cannot make any promises, I do intend to maintain it and add improvements if possible. Judging by the project histories on GitHub, however, AnyStyle has definitely been more active over the last... decade almost.

Two other projects I would suggest to evaluate are [grobid](https://github.com/kermitt2/grobid) and [science-parse](https://github.com/allenai/science-parse/). I have not used them myself, but they have been very active over the years.

**Comment by coryschires** [comment]
Hi @inukshuk,

I finally have an update regarding benchmarking which I hope you will find interesting. Before embarking on our own benchmarking, we were pleased to find this article on ArXiv: [Machine Learning vs. Rules and Out-of-the-Box vs. Retrained: An Evaluation of Open-Source Bibliographic Reference and Citation Parsers](https://arxiv.org/pdf/1802.01168.pdf). 

The authors evaluate ~12 different tools used to extract metadata from raw citation strings – including GROBID, ParsCit, and AnyStyle. Overall, the ML-based tool perform better than rule/regex-based tools (not surprising). According to their findings, AnyStyle unfortunately is less accurate that similar tools such as ParsCit and GROBID. That said, they benchmarked an old version of AnyStyle and much has changed (and presumably improved) in the interim. 

Based on this article, we have done some preliminary benchmarks comparing the latest versions of AnyStyle and GROBID. While it's too soon to say conclusively, GROBID does seem to be at least slightly more accurate. But, as you may expect, both libraries have inaccuracies, and sometimes AnyStyle is correct where GROBID is wrong and visa versa. Furthermore, GROBID only exports TEI which does not map perfectly to CSL-JSON (our desired output). 

For these reasons, I suspect the best solution will involve several libraries working together to achieve greater accuracy via some sort of consensus.

**Comment by a-fent** [comment]
Thank you @coryschires for the pointer to this useful paper. A few thoughts on reading it (and without any particular angle to promote)

- my reading of the result is that "state-of-the-art" tools (i.e. those underpinned by some kind of classification model, probably CRF - so including AnyStyle) perform similarly (and well) when they are trained on domain-specific data
- domain-specific training is especially important when dealing with idiosyncratic citation styles, such as the no-title style in the chemistry corpus examined in the paper (see bottom col.1 p5). I have observed that AnyStyle used on a mixed corpus does indeed struggle with this style.
- the field-specific precision and recall rates (table 3, p7) are usefully diagnostically for individual tools. There are tricky decisions separate from the training data, e.g. about how to tokenise (this is relevant for handling vol/issue) and how to decompose name strings into individual name-parts (this is an area where I think AnyStyle could use improvement).

Despite the high recall etc scores I wouldn't say that this whole problem is by any means solved - having tried all of the tools I have not yet found one capable of reliably making sense of footnote-citation styles common in German and French (perh. elsewhere) humanities and social sciences. Grobid sometimes identifies the German surname "zum Vergleich" ("see also") :)

**Comment by inukshuk** [comment]
Thanks @coryschires for the link to the paper (and the sleepless night it caused, ha!). I'm not quite sure what to make of it; the paper presents a good summary of some of the challenges involved and the its results reflect rather poorly on AnyStyle. At the same time, I feel like some of the statements about AnyStyle in particular are misleading and, most importantly, without seeing the training and control data sets it's hard to evaluate the results. What I will say is that version 0.x was definitely worse than the current release (in my own testing the error rate dropped by a third if not more); furthermore version 0.x was released with a very limited default model, consisting almost entirely of the data set used by ParsCit and was not intended to be used without adding your own data. This is also true for the current release, but the current default model is much better than the one we shipped with 0.x.

Without dwelling on it much further, I will say that from what I've seen and heard, I do believe that Grobid does a very good job. Since it uses the same CRF engine as AnyStyle, I don't think that one would be drastically better or worse than the other. Which one yields better results for you, I think, depends very much on your data and how well it aligns to the assumptions made during feature extraction and normalization (if any) by the parser authors. If you're curious to know, AnyStyle's latest development was based on theses, disserations and monographs across all departments at Stanford University so, out of the box, I think that the default model is best suited for similar input.

Since you mentioned consensus strategies, I would like to add that you can also use AnyStyle with different models (e.g., you could create different models specialized on different styles, languages, etc.) but what you would need in any case is a good way to measure the quality of results. If you happen to have a good idea how such measuring might work, I'd be interested to add that to the library because it could be used in many situations.

Thanks @a-fent for your thoughts, too. I agree with everything you said, and wanted to add that I've recently encountered [some references without title](https://github.com/inukshuk/anystyle/blob/master/res/parser/gold.xml#L12); there references I had to deal with, incidentally, included mainly journal abbreviations and, as result, I was able to get good results even using a model based on a mixed corpus. But in general, yes, missing titles might easily lead to worse results and using a dedicated model may make sense in such cases.

References in footnotes often follow different rules; that would definitely need some work to get good results.

As for parsing individual name parts, the reason why AnyStyle currently does that during normalization (i.e., in the second pass) is mainly in order to keep the training data simpler (it would be much harder to curate and maintain the data sets if we marked up all the name parts) -- I stand by that decision. However, the name parsing itself is currently purely deterministic and based mostly on the rules established for parsing names in BibTeX. It might be worthwhile to add a separate CRF model for parsing just the name segments. But I'd be interested to hear your ideas if you would like to see this area improved (probably best to discuss in a separate thread though).
